lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java

```diff
@@ -0,0 +1,169 @@
+import org.apache.reef.runtime.common.driver.api.*;
+import org.apache.reef.runtime.common.driver.resourcemanager.NodeDescriptorEvent;
+import org.apache.reef.runtime.common.driver.resourcemanager.ResourceAllocationEvent;
+import org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEvent;
+import org.apache.reef.runtime.common.driver.resourcemanager.RuntimeStatusEvent;
+import org.apache.reef.runtime.multi.client.parameters.SerializedRuntimeDefinitions;
+import org.apache.reef.runtime.multi.driver.org.apache.reef.runtime.multi.driver.parameters.RuntimeName;
+import org.apache.reef.runtime.multi.utils.RuntimeDefinition;
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.Injector;
+import org.apache.reef.tang.JavaConfigurationBuilder;
+import org.apache.reef.tang.Tang;
+import org.apache.reef.tang.annotations.Parameter;
+import org.apache.reef.tang.exceptions.InjectionException;
+import org.apache.reef.tang.formats.AvroConfigurationSerializer;
+import org.apache.reef.wake.EventHandler;
+import org.apache.reef.wake.time.runtime.event.RuntimeStart;
+import org.apache.reef.wake.time.runtime.event.RuntimeStop;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Hosts the actual runtime implementations and delegates invocations to them.
+ */
+final class RuntimesHost {
+  private final Set<String> runtimeDefinitions;
+  private Map<String, Runtime> runtimes;
+  private final Injector originalInjector;
+  private String defaultRuntimeName;
+
+  @Inject
+  private RuntimesHost(final Injector injector,
+                       @Parameter(SerializedRuntimeDefinitions.class) final Set<String> runtimeDefinitions) {
+    if (runtimeDefinitions == null || runtimeDefinitions.size() == 0) {
+      throw new RuntimeException("No runtime configurations are provided for multi-runtime");
+    }
+
+    this.runtimeDefinitions = runtimeDefinitions;
+    this.originalInjector = injector;
+  }
+
+  private synchronized void initialize() {
+    if (this.runtimes != null) {
+      return;
+    }
+
+    this.runtimes = new HashMap<>();
+    final AvroConfigurationSerializer serializer = new AvroConfigurationSerializer();
+
+    for (final String serializedRuntimeDefinition : runtimeDefinitions) {
+      Configuration config = null;
+      RuntimeDefinition rd = null;
+      try {
+        rd = parseSerializedRuntimeDefinition(serializedRuntimeDefinition);
+
+        if (rd.getDefaultConfiguration()) {
+          this.defaultRuntimeName = rd.getRuntimeName().toString();
+        }
+
+        config = serializer.fromString(rd.getSerializedConfiguration().toString());
+      } catch (IOException e) {
+        throw new RuntimeException("Unable to read runtime configuarion.", e);
+      }
+
+      final Injector rootInjector = Tang.Factory.getTang().newInjector();
+      try {
+        initializeInjector(rootInjector);
```

Comment:
We need new injector for each runtime as they define same named parameters which conflict with each other, so we need to isolate them in different injectors.